config/hostcredentials/transform.go

```diff
@@ -0,0 +1,76 @@
+// Package hostcredentials implements a TransformationProvider that fetches
+// credentials from the (oddly named) `host-secrets` service and replaces
+// objects of the form: {$hostcredentials: true} with the credentials.
+//
+// The configuration value `hostCredentialsUrls` must be a list of URLS for
+// host-secrets servers, which will be tried in order repeatedly until one
+// succeeds.
+//
+// Note that this transform will need to run before the "hostsecrets" transform.
+package hostcredentials
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/taskcluster/taskcluster-worker/config"
+)
+
+type provider struct{}
+
+func init() {
+	config.Register("hostcredentials", provider{})
+}
+
+func (provider) Transform(cfg map[string]interface{}) error {
+	urls, ok := cfg["hostCredentialsUrls"].([]interface{})
+	if !ok {
+		return errors.New(
+			"Expected 'hostCredentialsUrls' property to hold a list of URLs for host-secrets services")
+	}
+
+	var creds interface{}
+
+FetchLoop:
+	for {
+		for _, url := range urls {
+			log.Printf("Trying host-secrets server %s...", url)
+
+			resp, err := http.Get(url.(string))
+			if err != nil {
+				log.Printf("result: %s; continuing to next server", err)
+				continue
+			}
+
+			if resp.StatusCode != 200 {
+				log.Printf("result: %s; continuing to next server", resp.Status)
+				continue
+			}
+
+			body, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				log.Printf("result: %s; continuing to next server", err)
+				continue
+			}
+
+			err = json.Unmarshal(body, &creds)
+			if err != nil {
+				log.Printf("decoding JSON from server: %s; continuing to next server", err)
+				continue
+			}
```

Comment:
I can't imagine @jonasfj means the first one, since this is the reason we publish schemas, and we consistently validate inputs/outputs in all our systems, also between our own systems that talk with each other. I think gating inputs/outputs helps avoid nondeterministic behaviour, as code that accepts inputs will make assumptions about it, so if it hasn't validated these assumptions, it can lead to non-deterministic code or even crashes. I'm sure he meant the second option. :)